Update star difficulty bindables on mod settings change

frenzibyte opened the pull request:
Closes #9669 

Since there have been quite a lot of reports on this, I've decided to looking into it and it looks to be somewhat straightforward. I've separated the steps I went through to be able to support this each per commit.

I've had to apply a 100ms debounce for the beatmap difficulty cache to not be stuffed with update requests when a user starts dragging a mod setting, in a somewhat similar fashion to `AdvancedStats`.
 
https://user-images.githubusercontent.com/22781491/129656243-c337f261-8404-4123-987c-e9c32971883f.mp4

I think further improvements to this would be to have animation/counting support in `StarRatingDisplay` to feel more active with the difficulty change, but that can come along with #14128, as the current design cannot have such animation.

frenzibyte committed:
Add hash code support for `Mod`

frenzibyte committed:
Consider mod equality in difficulty cache lookup rather than acronym

frenzibyte committed:
Convert beatmap difficulty test into test scene and extend coverage

frenzibyte committed:
Track changes to mod settings in beatmap difficulty cache with 100ms debouncing

frenzibyte committed:
Refactor beatmap info wedge to not fully refresh on star difficulty change

Makes it look awkward when changing difficulty via mod settings for
example.

Now the changes should instead only affect the displayed components which consume it
directly.

frenzibyte committed:
Add visual test slider for changing star difficulty in beatmap wedge

peppy commented on osu.Game/Rulesets/Mods/Mod.cs:
```diff
@@ -197,6 +197,18 @@ public bool Equals(Mod other)
+
+            hashCode.Add(GetType());
+
+            foreach (var (_, prop) in this.GetSettingsSourceProperties())
+                hashCode.Add(ModUtils.GetSettingUnderlyingValue(prop.GetValue(this)));
```
I'm not sure we want to be doing this in `GetHashCode`. It is supposed to be a fast operation and this seems like it could have some overhead.

frenzibyte replied:
Good point, I think we can have a static cache for these as they return the properties of the type, nothing to do with the instance itself.

peppy replied:
Why are you implementing in `GetHashCode` in the first place though?

frenzibyte replied:
Because it is used in `DifficultyCacheLookup.GetHashCode()`, for equality comparisons during value retrieval I believe.

peppy replied:
Keep in mind `GetHashCode` is a best-effort return, not the ground truth. May want to read up on why it exists. May not be required in this case.

frenzibyte replied:
`MemoryCachingComponent`s check for existence of computed values by doing `cache.TryGetValue(lookup, out var value)`, and the implementation of that method seems to require hash code to function properly:
![CleanShot 2021-08-17 at 18 19 16](https://user-images.githubusercontent.com/22781491/129753902-2b181c80-a37a-4638-a31a-05223b120ae9.png)

(doesn't consider equality even when `Equals` would return true)

I'll go with caching the properties just so that `GetHashCode()` doesn't get hindered by the overhead.

bdach replied:
Caching the properties doesn't eliminate all overhead. There's still the cost of fetching the values via reflection.

Looking at this from a different angle, I'm not really sure whether difficulty values for mods with settings *should* be cached at all. The reason why I'm doubting that is that I'm a bit skeptical as to what the hit rate of such a cache would be, especially for things like difficulty adjust where you have some hundreds of admissible *values* for each property, let alone *combinations* of values.

In other words, what I'm trying to put under consideration here is *not* caching difficulty values if they come from a mod with non-standard setting values. That would sidestep the issue entirely.

frenzibyte replied:
It's not the value that's cached though? It's the property info itself, but yeah there's the overhead of fetching as well unfortunately.

Tempted to just leave this off but I'll need to benchmark because reflection is full of mysteries.

frenzibyte replied:
Yeah seems there's still an overhead from the fetching alone:

- hash `GetType()` only (no properties fetching):
    |      Method | Times |        Mean |      Error |     StdDev | Gen 0 | Gen 1 | Gen 2 | Allocated |
    |------------ |------ |------------:|-----------:|-----------:|------:|------:|------:|----------:|
    | ModHashCode |     1 |    19.62 ns |   0.128 ns |   0.120 ns |     - |     - |     - |         - |
    | ModHashCode |    10 |   181.62 ns |   2.087 ns |   1.629 ns |     - |     - |     - |         - |
    | ModHashCode |   100 | 2,982.04 ns | 126.045 ns | 369.668 ns |     - |     - |     - |         - |

- hash `GetType()` and fetched properties without caching:
    |      Method | Times |        Mean |     Error |    StdDev |      Median |    Gen 0 | Gen 1 | Gen 2 | Allocated |
    |------------ |------ |------------:|----------:|----------:|------------:|---------:|------:|------:|----------:|
    | ModHashCode |     1 |    55.11 us |  3.031 us |  8.842 us |    52.10 us |   2.5635 |     - |     - |      8 KB |
    | ModHashCode |    10 |   349.67 us | 29.659 us | 87.451 us |   295.46 us |  26.3672 |     - |     - |     82 KB |
    | ModHashCode |   100 | 2,729.53 us | 11.389 us |  8.891 us | 2,730.06 us | 265.6250 |     - |     - |    818 KB |

- hash `GetType()` and fetched properties with caching:
    |      Method | Times |       Mean |      Error |     StdDev |   Gen 0 | Gen 1 | Gen 2 | Allocated |
    |------------ |------ |-----------:|-----------:|-----------:|--------:|------:|------:|----------:|
    | ModHashCode |     1 |   6.074 us |  0.2061 us |  0.5946 us |  0.1526 |     - |     - |     520 B |
    | ModHashCode |    10 |  42.969 us |  0.8216 us |  2.3173 us |  1.6479 |     - |     - |   5,200 B |
    | ModHashCode |   100 | 500.425 us | 16.0374 us | 45.4956 us | 16.1133 |     - |     - |  52,000 B |

I'll look at improving this, by once-off retrieving all bindable settings rather than fetching them every equality comparsion call.

frenzibyte replied:
After caching all bindables from first retrieval, the results appear to be the following (`GetHashCode` with no properties added and `GetHashCode` with properties added, side to side):

|                  Method | Times |         Mean |      Error |       StdDev |       Median |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|------------------------ |------ |-------------:|-----------:|-------------:|-------------:|-------:|------:|------:|----------:|
| ModHashCodeNoProperties |     1 |     29.05 ns |   0.757 ns |     2.232 ns |     29.61 ns |      - |     - |     - |         - |
|             ModHashCode |     1 |    113.39 ns |   2.233 ns |     1.979 ns |    113.21 ns | 0.0203 |     - |     - |      64 B |
| ModHashCodeNoProperties |    10 |    359.92 ns |  25.999 ns |    75.014 ns |    339.14 ns |      - |     - |     - |         - |
|             ModHashCode |    10 |  2,262.31 ns | 159.937 ns |   464.006 ns |  2,101.73 ns | 0.2022 |     - |     - |     640 B |
| ModHashCodeNoProperties |   100 |  2,932.02 ns |  70.621 ns |   201.486 ns |  2,864.35 ns |      - |     - |     - |         - |
|             ModHashCode |   100 | 17,209.31 ns | 946.256 ns | 2,730.164 ns | 16,491.80 ns | 2.0294 |     - |     - |   6,400 B |

I'm still not sure what's causing the allocation when enumerating the settings and adding them to the hashcode.

peppy commented on osu.Game/Screens/Select/BeatmapInfoWedge.cs:
```diff
@@ -304,20 +293,22 @@ private void load(LocalisationManager localisation)
+                starDifficulty = difficultyCache.GetBindableDifficulty(beatmapInfo, (cancellationSource = new CancellationTokenSource()).Token);
+                starDifficulty.BindValueChanged(s =>
+                {
+                    difficultyColourBar.Current.Value = s.NewValue ?? default;
```
Why are these done here rather than direct binds?

frenzibyte replied:
`StarRatingDisplay` and `DifficultyColourBar` both accept type `StarDifficulty`, while the returned bindable is of type `StarDifficulty?`.

peppy replied:
Yet they handle weird defaults locally? Wouldn't it make more sense to either make `default` used everywhere or have them support nullability? Not sure.

frenzibyte replied:
Not sure either, `StarRatingDisplay` supporting nullability sounds wrong unless you're fine with making it display 0.00 or hide itself, I would expect a null value on a component to display something indicating its value.

peppy replied:
Well as it stands, you're just using `default`. Doesn't that give the same behaviour?

frenzibyte replied:
The `default` there won't actually be hit as the value change event doesn't invoke once immediately, I should've added a `Debug.Assert` instead.

peppy replied:
So the display elements have a hidden behaviour for when they don't have an initial value... which can only be reached once ever? That seems really weird.

frenzibyte replied:
Only for the difficulty colour bar which is currently visible, can hide that until a star difficulty is calculated then.

frenzibyte replied:
While I was going through making all components accept `StarDifficulty?` instead, it turns out the following will also be required to bind, unfortunately:

![CleanShot 2021-08-17 at 18 47 31](https://user-images.githubusercontent.com/22781491/129758441-59038aa2-9094-4862-a05e-c44d3e2b8710.png)

I'll change to an assert and leave the rest for another PR, as otherwise things will get out of scope given that `StarRatingDisplay` only accepts `StarDifficulty` itself.

frenzibyte committed:
Cache mod settings rather than fetching everytime

frenzibyte committed:
Add mod benchmark

frenzibyte committed:
Merge branch 'master' into mod-settings-difficulty-cache

frenzibyte committed:
Add animation support for the star rating display

frenzibyte committed:
Remove leftover unused using

frenzibyte committed:
Use animated star display in beatmap info wedge and synchronise bar

frenzibyte committed:
Merge spectrum test case

frenzibyte commented:
I've added animation support to the star rating display for the beatmap info wedge to feel more active when a user changes a difficulty setting:

https://user-images.githubusercontent.com/22781491/130006990-6e57e1ff-1a6d-4f67-b131-184188da773a.mp4

The BPM changing while the star rating display is still held until after 100ms feels a bit weird, makes me think about using `TransferValueOnCommit` on all mod settings sliders and call it a day.

peppy commented:
Spacing looks to have broken in the wedge:

![20210819 142800 (dotnet)](https://user-images.githubusercontent.com/191335/130013019-7b65746e-4b7b-4b8e-b49c-5b9f59b826bf.png)


Should probably be specified as part of the `FillFlowContainer`, not an individual component (https://github.com/ppy/osu/blob/25e6317e7f1a51dc6cd15edbafaccb10d499a4d1/osu.Game/Screens/Select/BeatmapInfoWedge.cs#L237)

frenzibyte committed:
Apply 5px vertical spacing on fill flow

Regressed, was margin { bottom = 5f } from the star rating display
creation method, which I've partly inlined.

frenzibyte commented:
Resolved, was an oversight from my end.

peppy committed:
Move some constants to `const`s

peppy commented:
I think this looks okay. @smoogipoo can you make sure you're okay with the hash/equality logic?

smoogipoo commented:
I looked at it before and it seemed reasonable and simple, but didn't comment because I'm not sure whether the discussion above was overcomplicating things too early.
